### PR TITLE
cpufreq plugin now can obtain CPU frequency on "broken" kernels (obse…

### PR DESCRIPTION
**Pull request by igsol**
cpufreq plugin now can obtain CPU frequency on "broken" kernels (observed in CentOS 7.1, 7.2) where "/sys/devices/system/cpu/cpuX/cpufreq/scaling_cur_freq" file is absent.

The fix relative to #624.

### CODE REVIEW

**Comment by rubenk** [comment]
Please keep the first line of your commit message below 80 characters (preferably below 50).

File: src/cpufreq.c [hunk 1]
```diff
@@ -18,28 +18,125 @@
...
+ * This file exposes not exactly the same thing as "scaling_cur_freq" does, but close by nature.
+ * More details: http://www.pantz.org/software/cpufreq/usingcpufreqonlinux.html
+ *
+ * Overall it is better to have a workaround for monitoring some CPU frequency on even with kernels that lack
+ * the aforementioned fix.
+ *
+ * Also, an optional parameter "Path" is introduced to specify alternative CPU frequency path.
+ * Example:
+ * <Plugin cpufreq>
+ *     Path "/sys/devices/system/cpu/cpu%d/cpufreq/cpuinfo_cur_freq"
```
Comment (rubenk) [inline src/cpufreq.c]: Please add this setting to the manpage (https://github.com/collectd/collectd/blob/master/src/collectd.conf.pod)

File: src/cpufreq.c [hunk 2]
```diff
@@ -18,28 +18,125 @@
...
 #include "collectd.h"
 #include "common.h"
 #include "plugin.h"
 
 #define MODULE_NAME "cpufreq"
+#define MAX_STR_L 256
 
 static int num_cpu = 0;
 
+static char const * freq_fname_def = "/sys/devices/system/cpu/cpu%d/cpufreq/"
```
Comment (rubenk) [inline src/cpufreq.c]: - This line break looks weird.
- Although both work, please use const char *freq_fname_def instead of char const * freq_fname_def.

File: src/cpufreq.c [hunk 3]
```diff
@@ -18,28 +18,125 @@
...
+									 "scaling_cur_freq";
+static char freq_fname[MAX_STR_L] = {0};
+
+static const char *config_keys[] =
+{
+	"Path"
+};
+static int config_keys_num = STATIC_ARRAY_SIZE (config_keys);
+
+int check_format (char const * src)
```
Comment (rubenk) [inline src/cpufreq.c]: const char *src

File: src/cpufreq.c [hunk 4]
```diff
@@ -18,28 +18,125 @@
...
+{
+	int pc = -2;
+	int d = -2;
+	char const * p;
```
Comment (rubenk) [inline src/cpufreq.c]: const char please.

File: src/cpufreq.c [hunk 5]
```diff
@@ -18,28 +18,125 @@
...
+			case 'd':
+				if (pc+1 == p-src) d=p-src;
+				break;
+		}
+	}
+
+	return pc+1==d;
+} /* check_format */
+
+static int cpufreq_config (char const * key, char const * value)
```
Comment (rubenk) [inline src/cpufreq.c]: const char please.

File: src/cpufreq.c [hunk 6]
```diff
@@ -18,28 +18,125 @@
...
+{
+	if (strcasecmp ("Path", key) == 0) {
+		if (check_format (value)) {
+			sstrncpy (freq_fname, value, sizeof(freq_fname));
+		}
+		else {
+			WARNING ("cpufreq: Path parameter is wrong: %s", value);
+			return -1;
```
Comment (rubenk) [inline src/cpufreq.c]: We prefer return (-1)

File: src/cpufreq.c [hunk 7]
```diff
@@ -18,28 +18,125 @@
...
+		}
+	}
+
+	return 0;
```
Comment (rubenk) [inline src/cpufreq.c]: We prefer return (0). In general, please keep the same style as the existing code in this file.

File: src/cpufreq.c [hunk 8]
```diff
@@ -18,28 +18,125 @@
...
+} /* cpufreq_config */
+
+static int setup_freq_fname ()
```
Comment (rubenk) [inline src/cpufreq.c]: Please use ansi c function definitions, so setup_freq_fname (void).

File: src/cpufreq.c [hunk 9]
```diff
@@ -53,9 +150,9 @@ static int cpufreq_init (void)
 			(num_cpu == 1) ? "" : "s");
 
 	if (num_cpu == 0)
-		plugin_unregister_read ("cpufreq");
+		return -1;
```
Comment (rubenk) [inline src/cpufreq.c]: Why this change? This seems unrelated to the issue you're fixing.

File: src/cpufreq.c [hunk 10]
```diff
@@ -53,9 +150,9 @@ static int cpufreq_init (void)
...
 
-	return (0);
+	return 0;
```
Comment (rubenk) [inline src/cpufreq.c]: Please keep the existing style.

File: src/cpufreq.c [hunk 11]
```diff
@@ -132,6 +227,8 @@ static int cpufreq_read (void)
 
 void module_register (void)
 {
+	plugin_register_config ("cpufreq", cpufreq_config,
```
Comment (rubenk) [inline src/cpufreq.c]: Please break this line properly.

**Comment by octo** [comment]
@igsol ping? Are you still working on this?

**Comment by igsol** [comment]
Sorry for quite long time of silence. 
I was too busy, then on vacation, terrible wi-fi, etc...
@octo No I am not working on this pull request right now. As you can see the review contains remarks regarding style only. Incorporation is a trivial. But I have no chance to make it until next week.

File: src/cpufreq.c (hunk 9 above)
Comment (igsol) [inline src/cpufreq.c]: 1) Please, look on https://github.com/collectd/collectd/pull/1730#discussion_r65807467
2) Because it's more correct way regarding collectd documentation.

**Comment by octo** [comment]
What keeps us from merging is primarily the missing documentation. There are two files that need to be touched: `src/collectd.conf.in` (from which the default config is generated) and `src/collectd.conf.pod` (from which the manpage is generated).

Since you appear to *intend* to eventually touch this up, I'll leave it open. I'd be grateful if you'd drop us a note should you abandon this change.

Best regards,
—octo

**Comment by igsol** [comment]
@octo No no, I wouldn't abandon this change. It's important for us to have cpufreq plugin functional on current CentOS. I think other also may prefer to have it OOB. 

I saw upper remark regarding collectd.conf.pod.
Now you mentioned collectd.conf.in. Do you know may be other places should be touched else?
It would be nice to have exhaustive info in advance and I'll provide final fix in upcoming weekend.
Does such term sound Ok for you?

**Comment by igsol** [comment]
Can anybody provide vim modeline actual in collectd project?

**Comment by octo** [comment]
Hi @igsol, no worries, take your time.

> Do you know may be other places should be touched else?

No, those are the two places where config options are documented.

Not sure what you mean by "vim modeline". That specific file is indented with a single tab. You can configure vim to use that with:

```vim
set sw=8 sts=8 noet
```

For new files, we prefer indentation with two spaces.

Best regards,
—octo

**Comment by igsol** [comment]
All what you mentioned in the review were done. I do home :)

**Comment by rubenk** [comment]
@igsol any chance you could rebase this onto master? I'd like to get this into 5.7.

**Comment by igsol** [comment]
@rubenk 
Hi Ruben,
Wow, when I committed my latest changes I didn't see any conflicts and the current problems of commit:"Merge remote-tracking branch 'origin/master' into cpufreq_alternative…".

It's not clear for me why it happened.

> any chance you could rebase this onto master? I'd like to get this into 5.7.
Sure, I'd glad, but need to know what the most correct and less problematic way to do it?
Can you please suggest what exactly I have to do with my changes?

**Comment by rubenk** [comment]
Hi @igsol 

> Sure, I'd glad, but need to know what the most correct and least problematic way to do it?
Can you please suggest what exactly I have to do with my changes?

I'm not sure I understand your questions. You just need to rebase on master and fix up the conflicts.

**Comment by igsol** [comment]
@rubenk 
Nevermind, I already found the good enough "howto": https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request
Rebasing in the progress, will finish soon.

**Comment by rubenk** [comment]
Perfect, thank you!

**Comment by rubenk** [comment]
I'm not sure what you did, but it didn't go well ;)
"igsol  wants to merge 341 commits into collectd:master from igsol:cpufreq_alternative_path".

That seems a bit much.

**Comment by igsol** [comment]
Yes, I see what I was merged wrong.
Fixing it already.

**Comment by igsol** [comment]
@rubenk 
I see "collectd.conf.5 has some POD errors!" in the upper buildbot output.
What I made wrong?
Locally (Ubuntu 16.04) the "make all" was finished w/o any errors.

**Comment by rubenk** [comment]
We have some extra checks that check if the POD has any errors.
Normally 'make all' swallows these.

In any case, I suggest you finish the rebase first, then I can help you track down that one error.

**Comment by igsol** [comment]
@rubenk I redo my early rebase changes from scratch.
 I've synced the master branch in my fork with upstream collectd/master, then made rebase of my commits.
Local build didn't show any errors.
Now Jenkins and buildbot again complain, but it seems the errors are not relative to my changes or am I missed something?
Can you please try to merge the pullrequest to master?

File: src/collectd.conf.in [hunk 12]
```diff
@@ -103,7 +103,7 @@
 #@BUILD_PLUGIN_CONNTRACK_TRUE@LoadPlugin conntrack
 #@BUILD_PLUGIN_CONTEXTSWITCH_TRUE@LoadPlugin contextswitch
 @BUILD_PLUGIN_CPU_TRUE@@BUILD_PLUGIN_CPU_TRUE@LoadPlugin cpu
-#@BUILD_PLUGIN_CPUFREQ_TRUE@LoadPlugin cpufreq
+@BUILD_PLUGIN_CPUFREQ_TRUE@LoadPlugin cpufreq
```
Comment (rubenk) [inline src/collectd.conf.in]: This enables the plugin by default. That's a change in behaviour and breaks on other platforms. Please remove it.

**Comment by rubenk** [comment]
Sorry, you still have conflicts.

**Comment by rubenk** [comment]
> I synced the original master into my forked master branch, then I made rebase.

I suspect that's why you're having difficulties. Your own master branch diverged. You'll need to skip that step and rebase on the original master.

File: src/collectd.conf.in (hunk 12 above)
Comment (igsol) [inline src/collectd.conf.in]: done

**Comment by igsol** [comment]
@rubenk 
> I suspect that's why you're having difficulties. Your own master branch diverged. You'll need to skip that step and rebase on the original master.

Yes, it was the case during my first rebase attempt (many conflicts occurred). After I realized it, I rolled back all my changed in the working branch, then I synced up the my master with original repository master, and finally redone rebase. Thus, right now it's all what I can make from my side. Sorry, I will not have time in the nearest days to devote it to conflicts resolutions.

**Comment by rubenk** [comment]
As you'll see on https://github.com/collectd/collectd/pull/1771/commits this PR contains two commits with the same title (1ceb7f76bb3f95fc5f719bcae14dcadf3e1eeef8 and 3fae41b026f8d8b0717742fbc00f8af899b03c50) + a merge from one of your branches.

The diff against master is 285 files changed, 7142 insertions(+), 18919 deletions(-).

I'm sorry but I'm not going to be able to apply this.

Please open a new PR with your changes against the master branch when you have the time.
